File: config/cas-mgmt-config-authz-ldap/src/main/java/org/apereo/cas/mgmt/config/CasManagementLdapAuthorizationConfiguration.java
package org.apereo.cas.mgmt.config;

import org.apereo.cas.authorization.LdapUserAttributesToRolesAuthorizationGenerator;
import org.apereo.cas.authorization.LdapUserGroupsToRolesAuthorizationGenerator;
import org.apereo.cas.configuration.CasConfigurationProperties;
import org.apereo.cas.configuration.CasManagementConfigurationProperties;
import org.apereo.cas.util.CollectionUtils;
import org.apereo.cas.util.LdapUtils;

import lombok.val;
import org.apache.commons.lang3.StringUtils;
import org.ldaptive.SearchOperation;
import org.pac4j.core.authorization.generator.AuthorizationGenerator;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.util.ArrayList;

/**
 * This is {@link CasManagementLdapAuthorizationConfiguration}.
 *
 * @author Misagh Moayyed
 * @since 5.0.0
 */
@Configuration(value = "casManagementLdapAuthorizationConfiguration", proxyBeanMethods = false)
@EnableConfigurationProperties({CasConfigurationProperties.class, CasManagementConfigurationProperties.class})
public class CasManagementLdapAuthorizationConfiguration {

    private static SearchOperation ldapAuthorizationGeneratorUserSearchExecutor(final CasManagementConfigurationProperties managementProperties) {
        val ldapAuthz = managementProperties.getLdap().getLdapAuthz();
        val operation = LdapUtils.newLdaptiveSearchOperation(ldapAuthz.getBaseDn(), ldapAuthz.getSearchFilter(),
            new ArrayList<>(0), CollectionUtils.wrap(ldapAuthz.getRoleAttribute()));
        operation.setConnectionFactory(LdapUtils.newLdaptiveConnectionFactory(managementProperties.getLdap()));
        return operation;
    }

    private static SearchOperation ldapAuthorizationGeneratorGroupSearchExecutor(final CasManagementConfigurationProperties managementProperties) {
        val ldapAuthz = managementProperties.getLdap().getLdapAuthz();
        SearchOperation operation = LdapUtils.newLdaptiveSearchOperation(ldapAuthz.getGroupBaseDn(), ldapAuthz.getGroupFilter(),
            new ArrayList<>(0), CollectionUtils.wrap(ldapAuthz.getGroupAttribute()));
        operation.setConnectionFactory(LdapUtils.newLdaptiveConnectionFactory(managementProperties.getLdap()));
        return operation;
    }

    @Bean
    public AuthorizationGenerator authorizationGenerator(final CasManagementConfigurationProperties managementProperties) {
        val ldapAuthz = managementProperties.getLdap().getLdapAuthz();

        if (StringUtils.isNotBlank(ldapAuthz.getGroupFilter()) && StringUtils.isNotBlank(ldapAuthz.getGroupAttribute())) {
            return new LdapUserGroupsToRolesAuthorizationGenerator(
                ldapAuthorizationGeneratorUserSearchExecutor(managementProperties),
                ldapAuthz.isAllowMultipleResults(),
                ldapAuthz.getGroupAttribute(),
                ldapAuthz.getGroupPrefix(),
                ldapAuthorizationGeneratorGroupSearchExecutor(managementProperties));
        }
        return new LdapUserAttributesToRolesAuthorizationGenerator(
            ldapAuthorizationGeneratorUserSearchExecutor(managementProperties),
            ldapAuthz.isAllowMultipleResults(),
            ldapAuthz.getRoleAttribute(),
            ldapAuthz.getRolePrefix());
    }
}

File: webapp-mgmt/cas-management-webapp-support-ldap/src/main/kotlin/org/apereo/cas/mgmt/config/CasManagementLdapAuthorizationConfiguration.kt
package org.apereo.cas.mgmt.config

import org.apache.commons.lang3.StringUtils
import org.apereo.cas.authorization.LdapUserAttributesToRolesAuthorizationGenerator
import org.apereo.cas.authorization.LdapUserGroupsToRolesAuthorizationGenerator
import org.apereo.cas.configuration.model.support.ldap.LdapAuthorizationProperties
import org.apereo.cas.util.CollectionUtils
import org.apereo.cas.util.LdapUtils
import org.ldaptive.ConnectionFactory
import org.ldaptive.SearchExecutor
import org.pac4j.core.authorization.generator.AuthorizationGenerator
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.cloud.context.config.annotation.RefreshScope
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.apereo.cas.configuration.CasConfigurationProperties
import java.util.ArrayList

/**
 * This is [CasManagementLdapAuthorizationConfiguration].
 *
 * @author Misagh Moayyed
 * @since 5.0.0
 */
@Configuration("casManagementLdapAuthorizationConfiguration")
@EnableConfigurationProperties(CasConfigurationProperties::class)
class CasManagementLdapAuthorizationConfiguration {

    @Autowired
    private val casProperties: CasConfigurationProperties? = null

    @RefreshScope
    @Bean
    fun authorizationGenerator(): AuthorizationGenerator<*> {
        val ldapAuthz = casProperties!!.mgmt.ldap.ldapAuthz
        val connectionFactory = LdapUtils.newLdaptivePooledConnectionFactory(casProperties.mgmt.ldap)

        return if (StringUtils.isNotBlank(ldapAuthz.groupFilter) && StringUtils.isNotBlank(ldapAuthz.groupAttribute)) {
            LdapUserGroupsToRolesAuthorizationGenerator(connectionFactory,
                    ldapAuthorizationGeneratorUserSearchExecutor(),
                    ldapAuthz.isAllowMultipleResults,
                    ldapAuthz.groupAttribute,
                    ldapAuthz.groupPrefix,
                    ldapAuthorizationGeneratorGroupSearchExecutor())
        } else LdapUserAttributesToRolesAuthorizationGenerator(connectionFactory,
                ldapAuthorizationGeneratorUserSearchExecutor(),
                ldapAuthz.isAllowMultipleResults,
                ldapAuthz.roleAttribute,
                ldapAuthz.rolePrefix)
    }

    private fun ldapAuthorizationGeneratorUserSearchExecutor(): SearchExecutor {
        val ldapAuthz = casProperties!!.mgmt.ldap.ldapAuthz
        return LdapUtils.newLdaptiveSearchExecutor(ldapAuthz.baseDn, ldapAuthz.searchFilter,
                ArrayList(0), CollectionUtils.wrap(ldapAuthz.roleAttribute))
    }

    private fun ldapAuthorizationGeneratorGroupSearchExecutor(): SearchExecutor {
        val ldapAuthz = casProperties!!.mgmt.ldap.ldapAuthz
        return LdapUtils.newLdaptiveSearchExecutor(ldapAuthz.groupBaseDn, ldapAuthz.groupFilter,
                ArrayList(0), CollectionUtils.wrap(ldapAuthz.groupAttribute))
    }
}
